NumLib/NewtonMethods/NewtonRaphson.h

```diff
@@ -0,0 +1,79 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef NUMLIB_NEWTONMETHODS_NEWTONRAPHSON_H_
+#define NUMLIB_NEWTONMETHODS_NEWTONRAPHSON_H_
+
+#include <logog/include/logog.hpp>
+
+#include <Eigen/Dense>
+
+namespace NumLib
+{
+
+/// Newton-Raphson solver for system of equations using an Eigen linear solvers
+/// library.
+/// The current implementation does not update the solution itself, but calls a
+/// function for the solution's update with the current increment.
+template <typename LinearSolver, typename JacobianMatrix,
+          typename JacobianMatrixUpdate, typename ResidualVector,
+          typename ResidualUpdate, typename SolutionUpdate>
+class Newton final
+{
+public:
+    Newton(LinearSolver& linear_solver, JacobianMatrixUpdate jacobian_update,
+           ResidualUpdate residual_update, SolutionUpdate solution_update,
+           int const maximum_iterations, double const tolerance_squared)
+        : _linear_solver(linear_solver),
+          _jacobian_update(jacobian_update),
+          _residual_update(residual_update),
+          _solution_update(solution_update),
+          _maximum_iterations(maximum_iterations),
+          _tolerance_squared(tolerance_squared)
+    {}
+
+    /// Returns number of iterations.
+    int solve(JacobianMatrix& jacobian, ResidualVector& residual) const
+    {
+        int iteration = 0;
+        ResidualVector increment;
+        do
+        {
+            _jacobian_update(jacobian);
+            _residual_update(residual);
+
+            if (residual.squaredNorm() < _tolerance_squared)
+                break;  // convergence criteria fulfilled.
+
+            increment = _linear_solver.compute(jacobian).solve(-residual);
+            //DBUG("Local linear solver accuracy |J dx - r| = %g",
+            //     (jacobian * increment + residual).norm());
+
+            _solution_update(increment);
+
+            //DBUG("Local Newton: Iteration #%d |dx| = %g, |r| = %g", iteration,
+            //     increment.norm(), residual.norm());
+        } while (iteration++ < _maximum_iterations);
+
+        return iteration;
```

Comment:
i think there is a problem if we return only `iteration`. If it is equal to the given max. iterations, it can mean the iteration is stopped because it reaches the max number (line 62) or is finished because it converged (line 52).